File: transportation-api/src/services/mock.py
"""Built-in mock data for offline / demo mode.

When ``settings.use_mock_data`` is enabled, the combined endpoint serves this
spec-conformant :class:`BusStopCombined` for any stop id instead of hitting
live providers. Kept as a standalone builder so tests can reuse it and the
route stays free of hardcoded test data.
"""
from __future__ import annotations

import datetime

from models import (
    BusStop,
    BusStopCombined,
    ETA,
    GeoPoint,
    Incident,
    MultilingualText,
    Weather,
    WeatherWarning,
)

_QUERY_TIME = datetime.datetime(2026, 7, 10, 8, 45, 0, tzinfo=datetime.timezone.utc)
_DATA_TS = datetime.datetime(2026, 7, 10, 8, 40, 0, tzinfo=datetime.timezone.utc)
_ETA_TS = datetime.datetime(2026, 7, 10, 8, 49, 0, tzinfo=datetime.timezone.utc)


def _build_mock_combined() -> BusStopCombined:
    stop = BusStop(
        id="946C74E30100FE80",
        name=MultilingualText(en="Cheung Sha Wan Plaza", tc="長沙灣廣場", sc="长沙湾广场"),
        location=GeoPoint(lat=22.333, lon=114.161),
        address=MultilingualText(en="Cheung Sha Wan Road, Kowloon", tc="九龍長沙灣道", sc="九龙长沙湾道"),
        routes=["1", "2", "6"],
        data_timestamp=_DATA_TS,
    )

    etas = [
        ETA(
            co="KMB", route="1", direction="O", service_type=1, seq=12,
            dest=MultilingualText(en="Central (Macao Ferry)", tc="中環（港澳碼頭）", sc="中环（港澳码头）"),
            eta_seq=1, eta=_ETA_TS, minutes_remaining=4,
            remark=MultilingualText(en="Scheduled", tc="預定", sc="预定"),
            data_timestamp=_DATA_TS,
        ),
        ETA(
            co="KMB", route="1", direction="O", service_type=1, seq=12,
            dest=MultilingualText(en="Central (Macao Ferry)", tc="中環（港澳碼頭）", sc="中环（港澳码头）"),
            eta_seq=2, eta=_ETA_TS + datetime.timedelta(minutes=11), minutes_remaining=15,
            remark=None, data_timestamp=_DATA_TS,
        ),
        ETA(
            co="KMB", route="2", direction="O", service_type=1, seq=8,
            dest=MultilingualText(en="West Kowloon Cultural District", tc="西九文化區", sc="西九文化区"),
            eta_seq=1, eta=_ETA_TS + datetime.timedelta(minutes=3), minutes_remaining=3,
            remark=MultilingualText(en="KMB staff on board", tc="九巴職員當值", sc="九巴职员当值"),
            data_timestamp=_DATA_TS,
        ),
    ]

    weather = Weather(
        description="Light Rain",
        temperature={"place": "Hong Kong Observatory", "value": 28, "unit": "C"},
        humidity={"value": 84, "unit": "%"},
        icon=[62],
        update_time=_DATA_TS,
        warnings=[
            WeatherWarning(
                code="WRAINA",
                title=MultilingualText(en="Amber Rainstorm Warning Signal", tc="黃色暴雨警告信號", sc="黄色暴雨警告信号"),
                severity="amber",
                contents="Amber Rainstorm Warning is in force. Rainfall exceeds 30 millimetres in an hour.",
                issue_time=_DATA_TS,
            )
        ],
        forecast=[{"date": "2026-07-11", "weather": "Rain", "max_temp": 30, "min_temp": 27}],
    )

    incidents = [
        Incident(
            id="TD20260710-00123",
            heading=MultilingualText(en="Road blocked due to accident", tc="因交通意外道路封閉", sc="因交通意外道路封闭"),
            detail=MultilingualText(en="One lane closed on Cheung Sha Wan Road.", tc="長沙灣道一條行車線封閉。", sc="长沙湾道一条行车线封闭。"),
            location=MultilingualText(en="Cheung Sha Wan Road", tc="長沙灣道", sc="长沙湾道"),
            district=MultilingualText(en="Sham Shui Po", tc="深水埗", sc="深水埗"),
            direction=None,
            road_type=MultilingualText(en="Road", tc="道路", sc="道路"),
            near_landmark=MultilingualText(en="Cheung Sha Wan Plaza", tc="長沙灣廣場", sc="长沙湾广场"),
            status=MultilingualText(en="ACTIVE", tc="生效中", sc="生效中"),
            announcement_date="2026-07-10 08:30",
            relevance="high",
        )
    ]

    return BusStopCombined(
        stop=stop, etas=etas, weather=weather, incidents=incidents, query_time=_QUERY_TIME
    )

File: web-app/src/services/mock.ts
import { EtaAggregate, SearchResponse } from './types';

/**
 * Mock data layer. Returns realistic data shapes that match the live
 * GET /api/v1/eta and GET /api/v1/search payloads, so the UI renders
 * meaningfully when the backend is down — or when VITE_USE_MOCK=true.
 */

export const MOCK_API_BASE = '__mock__';

const ROUTE_TERMINALS: Record<string, { en: string; tc: string; sc: string }> = {
  '1': { en: 'Star Ferry', tc: '尖沙咀碼頭', sc: '尖沙咀码头' },
  '10': { en: 'Kennedy Town', tc: '堅尼地城', sc: '坚尼地城' },
  '113': { en: 'Cross Harbour Tunnel', tc: '紅磡海底隧道', sc: '红磡海底隧道' },
  '11K': { en: 'Hung Hom Station', tc: '紅磡車站', sc: '红磡车站' },
};

const REMARKS: Record<string, { en: string; tc: string; sc: string }> = {
  on_time: { en: 'On time', tc: '準時', sc: '准时' },
  delayed: { en: 'Delayed', tc: '延誤', sc: '延误' },
  departed: {
    en: 'The final bus has departed from this stop',
    tc: '最後班次已過',
    sc: '最后班次已过',
  },
};

export function getMockEta(
  route: string,
  stop: string,
  lang: string
): EtaAggregate {
  const terminal =
    ROUTE_TERMINALS[route] || { en: 'Terminus', tc: '總站', sc: '总站' };
  const now = new Date();
  const mkEta = (mins: number | null, seq: number, remarkKey?: string) => {
    const etaTime = mins === null ? null : new Date(now.getTime() + mins * 60000);
    return {
      co: 'KMB',
      route,
      direction: 'O' as const,
      serviceType: 1,
      seq,
      dest: terminal,
      etaSeq: seq,
      eta: etaTime ? etaTime.toISOString() : '',
      minutesRemaining: mins,
      remark: remarkKey ? REMARKS[remarkKey] : null,
      dataTimestamp: now.toISOString(),
      status: mins === null ? ('scheduled' as const) : ('live' as const),
    };
  };

  const etas =
    route === '1'
      ? [mkEta(3, 1), mkEta(12, 2), mkEta(25, 3), mkEta(null, 4, 'departed')]
      : [mkEta(7, 1), mkEta(19, 2)];

  return {
    query: { route, stopId: stop, operator: 'KMB', lang },
    etas,
    weather: {
      temperature: { place: '香港天文台', value: 30, unit: 'C' },
      description: '天色良好',
      humidity: { value: 83, unit: 'percent' },
      icon: [75],
      updateTime: now.toISOString(),
      warnings: [
        {
          code: 'WTHU',
          title: {
            en: 'Thunderstorm Warning',
            tc: '雷暴警告',
            sc: '雷暴警告',
          },
          severity: 'red',
          contents: '雷暴警告現正生效，香港地區有幾陣狂風雷暴。',
          issueTime: now.toISOString(),
        },
        {
          code: 'WHOT',
          title: {
            en: 'Very Hot Weather Warning',
            tc: '酷熱天氣警告',
            sc: '酷热天气警告',
          },
          severity: 'warning',
          contents: '酷熱天氣警告現正生效。',
          issueTime: now.toISOString(),
        },
      ],
      forecast: null,
    },
    incidents: [
      {
        id: 'IN-MOCK-001',
        heading: { en: 'Road Incident', tc: '道路事故', sc: '道路事故' },
        detail: { en: 'Traffic Accident', tc: '交通意外', sc: '交通意外' },
        location: { en: 'Kwun Tong Road', tc: '觀塘道', sc: '觀塘道' },
        district: null,
        status: { en: 'NEW', tc: '最新情況', sc: '最新情况' },
        relevance: 'high',
        announcementDate: now.toISOString().slice(0, 16).replace('T', ' '),
        content: {
          en: 'Part of Kwun Tong Road closed due to traffic accident.',
          tc: '觀塘道因交通意外部分行車線封閉。',
          sc: '觀塘道因交通意外部分行车线封闭。',
        },
      },
    ],
    queryTime: now.toISOString(),
    degraded: false,
    mock: true,
  };
}

const MOCK_STOPS: Record<string, { en: string; tc: string; sc: string }> = {
  '946C74E30100FE80': { en: 'Cheung Sha Wan Plaza', tc: '長沙灣廣場', sc: '长沙湾广场' },
  'ABC1230000000001': { en: 'Central Station', tc: '中環站', sc: '中环站' },
  'ABC1230000000002': { en: 'Admiralty Station', tc: '金鐘站', sc: '金钟站' },
};

export function getMockSearch(q: string, lang: string): SearchResponse {
  const needle = q.trim().toLowerCase();
  if (!needle) {
    return { query: q, lang, total: 0, stops: [], routes: [] };
  }

  const routes = Object.keys(ROUTE_TERMINALS)
    .filter((r) => r.toLowerCase().includes(needle))
    .map((r) => ({
      id: r,
      kind: 'route' as const,
      operator: 'KMB',
      name: ROUTE_TERMINALS[r],
    }));

  const stops = Object.entries(MOCK_STOPS)
    .filter(([, n]) =>
      [n.en, n.tc, n.sc].some((t) => t.toLowerCase().includes(needle))
    )
    .map(([id, n]) => ({
      id,
      kind: 'stop' as const,
      name: n,
      routes: routes.map((r) => r.id),
    }));

  return {
    query: q,
    lang,
    total: routes.length + stops.length,
    stops,
    routes,
  };
}
